Emby.Server.Implementations/Session/SessionWebSocketListener.cs

```diff
@@ -99,5 +140,205 @@ private void EnsureController(SessionInfo session, IWebSocketConnection connecti
+                    // Start KeepAlive watcher
+                    var task = RepeatAsyncCallbackEvery(
+                        KeepAliveSockets,
+                        TimeSpan.FromSeconds(WebSocketLostTimeout * IntervalFactor),
+                        _keepAliveCancellationToken.Token);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Stops the KeepAlive watcher.
+        /// </summary>
+        private void StopKeepAlive()
+        {
+            lock (_keepAliveLock)
+            {
+                if (_keepAliveCancellationToken != null)
+                {
+                    _keepAliveCancellationToken.Cancel();
+                    _keepAliveCancellationToken = null;
+                }
+            }
+
+            lock (_webSocketsLock)
+            {
+                foreach (var webSocket in _webSockets)
+                {
+                    webSocket.Closed -= OnWebSocketClosed;
+                }
+                _webSockets.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Checks status of KeepAlive of WebSockets.
+        /// </summary>
+        private async Task KeepAliveSockets()
+        {
+            IEnumerable<IWebSocketConnection> inactive;
+            IEnumerable<IWebSocketConnection> lost;
+
+            lock (_webSocketsLock)
+            {
+                _logger.LogDebug("Watching {0} WebSockets.", _webSockets.Count);
+
+                inactive = _webSockets.Where(i =>
+                {
+                    var elapsed = (DateTime.UtcNow - i.LastKeepAliveDate).TotalSeconds;
+                    return (elapsed > WebSocketLostTimeout * ForceKeepAliveFactor) && (elapsed < WebSocketLostTimeout);
+                });
+                lost = _webSockets.Where(i => (DateTime.UtcNow - i.LastKeepAliveDate).TotalSeconds >= WebSocketLostTimeout);
+            }
+
+            if (inactive.Any())
+            {
+                _logger.LogInformation("Sending ForceKeepAlive message to {0} inactive WebSockets.", inactive.Count());
+            }
+
+            foreach (var webSocket in inactive)
+            {
+                try
+                {
+                    await SendForceKeepAlive(webSocket);
+                }
+                catch (WebSocketException exception)
+                {
+                    _logger.LogInformation(exception, "Error sending ForceKeepAlive message to WebSocket.");
+                    lost = lost.Append(webSocket);
+                }
+            }
+
+            lock (_webSocketsLock)
+            {
+                if (lost.Any())
+                {
+                    _logger.LogInformation("Lost {0} WebSockets.", lost.Count());
+                    foreach (var webSocket in lost.ToList())
+                    {
+                        // TODO: handle session relative to the lost webSocket
+                        RemoveWebSocket(webSocket);
```

Comment:
What if socket became "non-lost" after you had computed `lost` and `elapsed` but before we got here?
Maybe we should `lock (_webSocketsLock)` the whole function?

Reply:
Not knowing how long it takes for `SendForceKeepAlive` to complete, I've split the lock to avoid keeping onto it for too long, mainly because there's a `foreach` that awaits for all the sockets that need to be contacted.

Also, I had in mind this strict requirement: a client either keeps its time under `WebSocketLostTimeout` or it's considered to be lost.

So, with this rule in mind, using or not a lock after computing the `lost` list won't change the end result for those WebSockets. For the `elapsed` list what happens is that an extra `ForceKeepAlive` message will be sent, even when the client recently updated its status.

To be less restrictive, I could move the `lost = _webSockets.Where...` line in the second lock and merge it with the WebSockets lost from `inactive`.

Was this your concern?